src/backend/lang_c/c_simple_ast.rs

```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
         let mut converter = CASTConverter::new(&self);
         converter.to_c_ast()
     }
+
+}
+
+/// SimpleCAST should meet following conditions.
+/// 1. There are at most 1 ActionEdge::Normal from each node.
+/// 2. There are ActionEdge::IfThen, ValueEdge::Conditional from ActionNode::If
+/// 3. The targets of value edges are ValueNode, The target of action edges are ActionNode.
+/// 4. The destination node of GotoDst edge is ActionNode.
+pub struct SimpleCASTVerifier {
 }
 
+type Verifier = Fn(NodeIndex, &SimpleCAST) -> Result<(), String>;
+impl SimpleCASTVerifier {
+    pub fn verify(cast: &SimpleCAST) -> Result<(), String> {
+        Self::verify_each_node(cast, &Self::verify_normal_action, "Normal action")?;
+        Self::verify_each_node(cast, &Self::verify_if, "If")?;
+        Self::verify_each_node(cast, &Self::verify_edge_action, "Edge-Action")?;
+        Self::verify_each_node(cast, &Self::verify_goto, "Goto")?;
+        Ok(())
+    }
+
+    fn verify_each_node(cast: &SimpleCAST, verifier: &Verifier, name: &str) -> Result<(), String> {
+        let mut is_valid = true;
+        let mut ret = String::new();
+        let nodes = cast.ast.node_indices();
+        for node in nodes {
+            if let Err(msg) = verifier(node, cast) {
+                ret = format!("{}; {}", ret, msg);
+                is_valid = false;
+            }
+        }
+        if is_valid {
+            Ok(())
+        } else {
+            Err(format!("{} @ {}", ret, name.to_string()))
+        }
+    }
+
+    // 1. There are at most 1 ActionEdge::Normal from each node.
+    fn verify_normal_action(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(_)) => {},
+            _ => return Ok(()),
+        };
+        let normal_actions = cast.ast.edges_directed(node, Direction::Outgoing)
+            .into_iter()
+            .filter(|e| {
+                match e.weight() {
+                    &SimpleCASTEdge::Action(ActionEdge::Normal) => true,
+                    _ => false,
+                }
+            }).collect::<Vec<_>>();
+        if normal_actions.len() <= 1 {
+            Ok(())
+        } else {
+            Err(format!("There are {} next normal actions @ {:?}",
+                        normal_actions.len(), node))
+        }
+    }
+
+    // 2. There are ActionEdge::IfThen, ValueEdge::Conditional from ActionNode::If
+    fn verify_if(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(ActionNode::If)) => {},
+            _ => return Ok(()),
+        };
+        let mut ret = String::new();
+        let cond = cast.branch_condition(node);
+        if cond.is_none() {
+            ret = format!("{}; No condition node is found.", ret);
```

Comment:
`ret` is empty before here, so after this statement `ret` will start with `"; ..."`. I don't think that's what you intended.